### PR TITLE
[LLP] Mutable contract state cache uses Offset as cache_index [DPP-1015]

### PR DESCRIPTION
**Pull request by tudor-da**

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionsReader.scala [hunk 1]
```diff
@@ -444,11 +446,17 @@ private[dao] final class TransactionsReader(
         )
       }
       .map(event => (event.eventOffset, event.eventSequentialId) -> event)
-      .mapMaterializedValue(_ => NotUsed)
+
+    val groupedByOffset = TransactionsReader
+      .groupContiguous(contractStateEventsSource)(by = { case ((offset, _), _) => offset })
```
Comment (tudor-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/dao/events/TransactionsReader.scala:451]: Other streams are grouped by `transactionId`. Does anyone see any problem of grouping here by `offset` instead?

**Review by nmarton-da — COMMENTED** [review]
It seems to me that changing to batched update/offset approach with the current MCSC has some non-trivial repercussions. Sorry for me asking too late: if we want to improve on the MCSC to turn that into solution which removes the lock contention, then doing these changes to Cache, StateCache and MCSC AFAICT would take different turns (many pieces in this PR would be not necessary).
I am especially wondering from the aspect of risk what we take with this change.

I made some comments: not sure about the mechanics.

File: ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala [hunk 2]
```diff
@@ -235,62 +235,59 @@ private[platform] class MutableCacheBackedContractStore(
...
+            keyMappingsBuilder.addOne(key -> Unassigned)
+          }
+          contractMappingsBuilder.addOne(archived.contractId, Archived(archived.stakeholders))
+        case _: LedgerEndMarker => ()
+      }
+
+      val keyMappings = keyMappingsBuilder.result()
+      val contractMappings = contractMappingsBuilder.result()
+
+      val validAt = eventsBatch.head.eventOffset
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala:282]: All items in this batch have the same offset, as guaranteed by the getContractStateEvents

File: ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala [hunk 3]
```diff
@@ -235,62 +235,59 @@ private[platform] class MutableCacheBackedContractStore(
...
 
   private def nonEmptyIntersection[T](one: Set[T], other: Set[T]): Boolean =
     one.intersect(other).nonEmpty
 
-  private def updateOffsets(event: ContractStateEvent): Unit = {
-    metrics.daml.execution.cache.indexSequentialId
-      .updateValue(event.eventSequentialId)
-    event match {
+  private def updateOffsets(eventsBatch: Seq[ContractStateEvent]): Unit =
+    eventsBatch.foreach {
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala:247]: Nothing changed in the frequency of LedgerEndMarker. There should be only one every polled ledger end.

File: ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala]: Amended the implementation to support multi-transaction updates: [367dc5c](https://github.com/digital-asset/daml/pull/13733/commits/367dc5cd7495329a5882c8f579fce713df41c17e)

**Review by nmarton-da — APPROVED** [review]
Since the DB lookups changed from event-seq-id to offset, this needs to be benchmarked in LR (although not sure any effects would be observable because of the MCSC)

**Pull request merged**